gcs_light_client.mediawiki

```diff
@@ -0,0 +1,1056 @@
+
+      // Otherwise, if the filter is empty (created from a block with a single
+      // coinbase transaction whose output script contains no push datas), then
+      // we'll hash the zero_hash.
+      case None:
+          double-sha-256(filter_header(n-1) || double-sha-256(zero_hash))
+
+   where filter(n) is the filter for block height n
+</pre>
+
+The filter header for the genesis block uses the hash stored in the prevblock
+field of the genesis block header itself, as there's no prior filter header
+(by definition).
+
+Due to the nature of filter construction, it's possible to construct a block
+such that an "empty" filter will be produced. This is the case of a coinbase
+transaction that has no data pushes in its output script. In this case, the
+"hash" of said filter is simply "32 zeroes". 
+
+
+We now introduce two new messages to support the fetching and verification of
+the filter header chain by light clients. 
+
+The <code>getcfheaders</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumBlockLocators
+| uint64
+| Number of block locators.
+|-
+| NumBlockLocators * 32
+| BlockLocatorHashes
+| [NumBlockLocators][32]byte
+| Block locator hashes, with the same semantics as in <code>getheaders</code>.
+|-
+| 32
+| HashStop
+| [32]byte
+| Hash to stop at.
+|-
+| 1
+| FilterType
+| byte
+| Type of filter header being requested.
+|}
+
+The <code>BlockLocators</code> within the message are to be interpreted
+identically to the <code>BlockLocators</code> within Bitcoin's
+<code>getheaders</code> and <code>getblocks</code> messages <ref>https://en.bitcoin.it/wiki/Protocol_documentation</ref>.
+
+The <code>cfheaders></code> message MUST be sent in response to a
+<code>getcfheaders</code> message for a particular block hash. The
+<code>cfheaders</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Descriptions
+! Data Type
+! Comments
+|-
+| 32
+| StopHash
+| []byte
+| Block hash for the last filter header returned, for locating the filter headers in the blockchain.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter headers being returned.
+|-
+| Var-Int
+| NumHeaders
+| uint64
+| Hash to stop at.
+|-
+| NumHeaders * 32
+| HeaderHashes
+| [NumHeaders][32]byte
+| Slice of filter headers.
+|}
+
+=== Compact Filters ===
+
+The last set of messages we introduce are for ''fetching'' the compact filters
+themselves. Light clients can SHOULD use these two messages to request a
+compact filter for a particular block hash.
+
+The <code>getcfilter</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| 32
+| BlockHash
+| [32]byte
+| Block hash of the Bitcoin block for which the client wishes to fetch a filter.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter requested.
+|}
+
+The <code>cfilter</code> message MUST be sent in response to a
```

Comment:
This is too restrictive. Need to specify what the restrictions are on the blockhash field (eg that it is something for which you've recently received a headers message from the peer for or so). Might also consdier clarifying that you should respond with the data for the same block and filter type  as the request :p.